onnxruntime/core/providers/webnn/webnn_execution_provider.cc

```diff
@@ -10,13 +10,101 @@
 #include "core/graph/graph_viewer.h"
 #include "core/session/onnxruntime_cxx_api.h"
 #include "core/common/safeint.h"
+#include "core/common/inlined_containers.h"
+#include "core/common/hash_combine.h"
 
 #include "builders/model.h"
 #include "builders/helper.h"
 #include "builders/model_builder.h"
 
 namespace onnxruntime {
 
+struct WebNNContextOptions {
+  std::optional<std::string> device_type;
+  std::optional<int> num_threads;
+  std::optional<std::string> power_preference;
+
+  [[nodiscard]] emscripten::val AsVal() const {
+    emscripten::val options = emscripten::val::object();
+    if (device_type.has_value()) {
+      options.set("deviceType", device_type.value());
+    }
+    if (num_threads.has_value()) {
+      options.set("numThreads", num_threads.value());
+    }
+    if (power_preference.has_value()) {
+      options.set("powerPreference", power_preference.value());
+    }
+    return options;
+  }
+
+  bool operator==(const WebNNContextOptions& other) const {
+    return std::tie(device_type, num_threads, power_preference) ==
+           std::tie(other.device_type, other.num_threads, other.power_preference);
+  }
+};
+
+}  // namespace onnxruntime
+
+// Specialize std::hash for WebNNContextOptions.
+template <>
+struct std::hash<::onnxruntime::WebNNContextOptions> {
+  size_t operator()(const ::onnxruntime::WebNNContextOptions& options) const {
+    size_t hash{0xbc9f1d34};  // seed
+    if (options.device_type.has_value()) {
+      onnxruntime::HashCombine(options.device_type.value(), hash);
+    }
+    if (options.num_threads.has_value()) {
+      onnxruntime::HashCombine(options.num_threads.value(), hash);
+    }
+    if (options.power_preference.has_value()) {
+      onnxruntime::HashCombine(options.power_preference.value(), hash);
+    }
+    return hash;
+  }
+};
+
+namespace onnxruntime {
```

Comment:
It is not duplicated, the onnxruntime namespace ends on line 47. This is because it template specialization 
for `std::hash<::onnxruntime::WebNNContextOptions>` must be declared inside the `std` namespace and before it is used on line 107 by `InlinedHashMap<WebNNContextOptions, emscripten::val> contexts_;`.